rebasebot/bot.py

```diff
@@ -170,6 +172,70 @@ def _in_excluded_commits(sha: str, exclude_commits: list) -> bool:
     return False
 
 
+def _find_last_rebase_merge_commit(gitwd: git.Repo, source_repo: Repository, ancestry_path_merges) -> Commit:
+    logging.info("Searching for merge commit from previous rebasebot run to identify downstream commits")
+    pygit2_repo = pygit2.Repository(gitwd.working_tree_dir)
+    for merge_line in ancestry_path_merges:
+        sha, commit_message, committer_email = merge_line.split(" || ", 2)
+        logging.info(f"Checking: \"{commit_message}\"")
+        if "openshift-merge-bot[bot]@users.noreply.github.com" in committer_email:
+            logging.info("Skipping this merge as this bot only merges downstream")
+            # This skip is not required and is here only to improve performance.
+            continue
+        merge = gitwd.commit(sha)
+
+        # Now we get both parents of the merge commit and check if one of them is an ancestor of any
+        # of the upstream branches. If it is, we know that this merge commit is the last rebase merge commit
+        for parent in merge.parents:
+            for upstream_branch in source_repo.branches():
+                # When merge base is the same as parent, it means that the parent is an ancestor of the upstream branch
+                merge_base = pygit2_repo.merge_base(parent.hexsha, upstream_branch.commit.sha)
+                if merge_base is not None and merge_base.hex == parent.hexsha:
+                    logging.info("Found merge commit from previous rebase: %s", sha)
+                    logging.info("Its parent %s is ancestor of %s", parent.hexsha, upstream_branch.name)
+                    return merge
+    return None
+
+
+def _identify_downstream_commits(gitwd: git.Repo, source: GitHubBranch, dest: GitHubBranch,
+                                 source_repo: Repository) -> str:
+    # Merge base is the last shared commit of source branch and destination branch
+    merge_base = gitwd.git.merge_base(f"source/{source.branch}", f"dest/{dest.branch}")
+    logging.info(f"Merge base of source/{source.branch} and dest/{dest.branch}: %s", merge_base)
+
+    # ancestry_path_merges are merge commits on ancestry path from merge base to destination branch
+    ancestry_path_merges = gitwd.git.log("--pretty=format:%H || %s || %aE", "--ancestry-path", "-r", "--merges",
+                                         f"{merge_base}..dest/{dest.branch}").splitlines()
+
+    logging.info(f"""Merges on ancestry-path from merge_base=({merge_base}) to dest/{dest.branch} branch:"
```

Comment:
This throws an invalid interpolation for me.

This works instead:
```suggestion
    val = '\\n'.join(ancestry_path_merges)
    logging.info(f"Merges on ancestry-path from merge_base=({merge_base}) to dest/{dest.branch} branch: {val}")
```